### ISSUE TITLE
Add ability to proxy local domain

### ISSUE THREAD

**Issue by jeromecoupe** [11ty/eleventy-dev-server#9]
In the spirit of using this outside of 11ty / Jamstack contexts, a Browsersync feature I am missing righ now (unless I am mistaken) is the ability to proxy a local domain.

**Comment by zachleat** [comment]
https://browsersync.io/docs/options/#option-proxy this one, right?

**Comment by jeromecoupe** [comment]
Indeed. I think having such an option opens up a lots of possibilities for a local dev server. Prime use case for me is when  working with local installs of more traditional CMSes, etc. With a proxy and a watch dir(s) option, I watch my templates, assets dirs, etc. and proxy my local domain.

**Comment by jeromecoupe** [comment]
This is (mainly) related to uses cases for Browsersync outside of an Eleventy context, so maybe it is oout of scope for this project. I work a lot with Craft CMS in headless mode or not. In those instances I run a LAMP stack locally that mirrors my production server as closely as possible. A custom `.test` domain is mapped to each site: `https://craft.sandbox.test` for example.

I then run Browsersync in proxy mode through the CLI in NPM script so that `http://localhost:3000` for example proxies `https://craft.sandbox.test`.

I can then watch any files / directories I want with Browsersync 

Sample NPM scripts (sorry no public repo right now)

- ./web/ is my webroot

```
"scripts": {
    "clean": "rm -rf './web/dist'",
    "styles:dev": "sass --embed-source-map --source-map-urls=absolute ./src/scss/main.scss ./web/dist/css/main.css",
    "styles:prod": "sass --no-source-map --style=compressed ./src/scss/main.scss ./web/dist/css/main.css",
    "poststyles:prod": "postcss ./web/dist/css/main.css --replace --no-map --use autoprefixer cssnano",
    "copy:images": "copyfiles -u 1 './src/img/**/*' './web/dist'",
    "copy:fonts": "copyfiles -u 1 './src/fonts/**/*' './web/dist'",
    "copy": "npm-run-all --parallel copy:images copy:fonts",
    "scripts:prod": "esbuild ./src/js/main.js --target=es6 --bundle --minify --outfile=./web/dist/js/main.bundle.js",
    "scripts:dev": "esbuild ./src/js/main.js --target=es6 --bundle --outfile=./web/dist/js/main.bundle.js",
    "cachebust": "node build_scripts/cachebust.js",
    "watch:scripts": "onchange './src/js/**/*' -- npm run scripts:dev",
    "watch:styles": "onchange './src/scss/**/*' -- npm run styles:dev",
    "build": "npm-run-all clean --parallel copy cachebust styles:prod scripts:prod",
    "server": "browser-sync start --proxy 'nisabalgg.craft.test' --serveStatic './web/dist' --files './web/dist' './templates' './translations' --no-open",
    "dev": "npm-run-all --parallel server watch:*"
  },
```

**Comment by chalkygames123** [comment]
@zachleat For example, we may want another server to interpret PHP while taking advantage of the dev server's features, and Browsersync can easily achieve this with a command like `php -S localhost:8000` and `npx browser-sync start --proxy 'localhost:8000'` (or `npx browser-sync 'http://localhost:8000'` for short). It would be great if the same could be done with the Eleventy Dev Server.

**Comment by pfist** [comment]
> Can you provide one (or more 😅) very specific example of how you use this? I just want to make sure I understand exactly what you’re looking for

I can. I use [Ghost]() for most of my websites, and I use a local install to develop themes. Ghost runs on `locahost:2368`, which requires a proxy to work with dev servers. If I run eleventy-dev-server at the same time as Ghost, it prevents Ghost from working, and all routes return a `Cannot GET` error in the browser. With BrowserSync, I simply add Ghost's domain as a proxy:

```js
browserSync.init({proxy: 'http://localhost:2368'})
```

It would be great to have this in eleventy-dev-server. I love how fast and lightweight it is compared to BrowserSync, but right now I can't use it for Ghost theme development.

**Comment by castastrophe** [comment]
> Can you provide one (or more 😅) very specific example of how you use this? I just want to make sure I understand exactly what you’re looking for

I would find this very helpful for our (@adobe/spectrum-css) documentation site which is migrating to 11ty. When in development mode, I'd like my developers to be able to live update the docs site and run the storybook server in parallel - with our built-in preview link in the navigation linking directly to the proxy instead of a static storybook build.

**Comment by KonnorRogers** [comment]
Just chiming in, I know its undocumented, but would a workaround for now be to add a middleware to the 11ty dev server?

I went spelunking through the Vite server and found this:

https://github.com/11ty/eleventy-plugin-vite/blob/fa4708b01fd0a660189c098ac02aa57e75d19da9/.eleventy.js#L26-L40

and am semi-curious if it could work for proxying as well.

**Comment by zachleat** [comment]
This is possible today using the `onRequest` option: https://www.11ty.dev/docs/dev-server/#on-request-for-request-time-processing

```js
eleventyConfig.setServerOptions({
	onRequest: {
		// proxy all paths (change this!)
		"/*": async function(req) {
			let response = await fetch(new URL(req.url.pathname, "https://www.zachleat.com/"));

			return {
				status: response.status,
				headers: {
					"content-type": response.headers.get("content-type"),
				},
				body: await response.bytes(),
			};
		}
	}
});
```

Further, I just added a feature in the soon-to-be-released Dev Server 2.0.7 to support `Response` instances returned in `onRequest` callbacks, so the above will be simplified to this:

```js
eleventyConfig.setServerOptions({
	onRequest: {
		// proxy all paths (change this!)
		"/*": function(req) {
			return fetch(new URL(req.url.pathname, "https://www.zachleat.com/"));
		}
	}
});
```

**Comment by zachleat** [comment]
Just one more small note: `onRequest` originally shipped in Dev Server v2.0.0 so it should be available to all Eleventy Core v3.0.0 installs: https://github.com/11ty/eleventy-dev-server/releases/tag/v2.0.0
